keps/sig-node/xxx-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1368 @@
+interface then made such local resources available to containers. But
+for many newer devices, this approach and the Kubernetes API for
+requesting these custom resources are too limited:
+
+- *Device initialization*: When starting a workload I’d like to have
+  the device reconfigured or reprogrammed by the orchestration. For
+  security reasons workloads should not be able to reconfigure devices
+  directly.
+
+  *Limitation*: Currently, it’s impossible to specify the desired
+  device properties that are required for reconfiguring devices.
+
+- *Device cleanup*: When my workload is finished, I would like to have
+  a mechanism for cleanup of the device, that will ensure that device
+  does not contain traces/parameters/data from previous workloads and
+  appropriate power state/shutdown.
+
+  *Limitation*: Post-stop actions are not supported.
+
+- *Partial allocation*: When deploying a container I’d like to be able
+  to use part of the shareable device inside a container and other
+  containers should be able to use other free resources on the same
+  device.
+
+  *Limitation*: Current implementation of the device plugin doesn’t
+  allow to allocate part of the device.
+
+- *Optional allocation*: When deploying workload I’d like to specify
+  soft(optional) device requirements. If the device exists and it’s
+  allocatable it would be allocated. If not - workload will be run on
+  the node without a device. GPU and crypto-offload engines are
+  examples of this kind of device. If they’re not available, workloads
+  can still run by falling back to using only the CPU for the same
+  task.
+
+  *Limitation*: Optional allocation is supported neither by the device
+  plugins nor by current Pod resource declaration.
+
+- *Support Over the Fabric devices*: When deploying a container, I’d
+  like to utilize devices available over the Fabric (PCIe, CXL,
+  Network, special links, etc).
+
+  *Limitation*: Device Plugins framework supports only local devices.
+
+- *Access to the container image*: When deploying a container that
+  needs access to a GPU device, I would like my container to
+  gracefully fail to start (rather than SIGSEGV at runtime) if a
+  minimum driver requirement is not met on the host.
+
+  *Limitation*: GPU driver requirements are currently stored as labels
+  on the container image, and device plugins do not have access to the
+  container image.
+
+- *Perform container runtime specific operations*: When deploying a container
+  that needs access to a device, I would like to be able to reuse the
+  same pod spec, irrespective of the underlying container runtime in
+  use (e.g. kata vs. runc).
+
+  *Limitation*: At present, there is no way to perform
+  runtime-specific operations that may need to occur as a result of
+  injecting a device into a container (device plugins are runtime
+  agnostic). A good example is supporting GPU passthrough
+  virtualization on kata vs. runc.
```

Comment:
What runtime-specific operations are needed? I do not understand what is meant by "GPU passthrough  virtualization on kata vs. runc". 
"

Reply:
@kad, @bart0sh: can you clarify?

Reply:
typically device passthrough to a VM sandbox (e.g., created by kata-containers container runtime) takes the GPU device bound to `/dev/vfio/xyx` and passes it to VMM. this device shows up as `/dev/dri/cardX` in the VM for the container to use.

In case of runc, the device on the host is `/dev/dri/cardX` which is added to the container.

if a container wants "gpu" resource, CDI needs to know if that is consumed by kata or runc to prepare the device correctly

Reply:
@cathyhongzhang as @mythi  mentioned, device passtrough should be handled by hypervisor of VM based runtimes as example. It might be more complex cases, e.g. for FPGA to perform configuration of device via management port on host side and to VM provide access only to "user" port device node.

Reply:
Is that covered by CDI internally or is that something that we need to take care of in this KEP?

Right now, both allocation and the  `NodePrepareResource` call are agnostic about the runtime that will be used.

This also makes it questionable whether a resource can be used by more than one pod. What if one pod uses runc and the other kata?

Reply:
runtime side CDI spec (json) specifies host devices (nodes, permissions, etc)... This is will become part of OCI spec that containerd/cri-o using to communicate with runc/kata/others. This KEP scope is up to creating those CDI json files. The rest is out of scope.

Reply:
So can we remove this particular item from the KEP?

Reply:
no, this scenario described in this paragraph is still valid: the cdi.json file might include instructions that potentially can be runtime specific. we had discussion this week COD meeting where Evan mentioned that potentially some conditional fields are needed in CDI spec, thus this paragraph is just sub-case of conditional parameters injection.

Reply:
So it is relevant for CDI, but not for dynamic resource allocation, right? This KEP isn't about what the plugins put into the JSON files.

We probably should make it clearer which of these items are the motivation for CDI and which of them motivate specific design choices in this KEP. For a reader of this KEP it will be confusing to see limitations listed which then don't get addressed as part of the KEP.

Reply:
I have re-arranged the "motivation" section in https://github.com/pohly/enhancements/pull/4